File: clients/dart/lib/src/transliterate/transliterate.dart
import '../phase.dart';



/// A engine for running the phases of transliteration, similar to the ICU4J transliterator.
abstract class Transliterate {

  final String name;  // For identification
  late List<Phase> _transliteratePhases;

  Transliterate( this.name) {
    _transliteratePhases = [];
  }
  Phase addPhase() {
    Phase newPhase =  Phase();
    newPhase.setInfo(' ${_transliteratePhases.length}');
    _transliteratePhases.add(newPhase);
    return newPhase;
  }


  String convert(String inString) {
    return _runAllPhases(inString);
  }

  String _runAllPhases(String inString) {
    String outString = inString;

    for (final phase in _transliteratePhases) {
      outString = phase.runPhase(outString);
    }
    return outString;
  }

}

File: clients/c#/MyanmarTools/Transliterate/Transliterate.cs
using System.Collections.Generic;

namespace MyanmarTools.Transliterate
{
    /// <summary>
    /// A engine for running the phases of transliteration, similar to the ICU4J transliterator.
    /// </summary>
    public abstract class Transliterate
    {

        public string Name { get; private set; }  // For identification
        private List<Phase> TransliteratePhases = new List<Phase>();


        public Transliterate(string Name)
        {
            this.Name = Name;
        }
        public Phase AddPhase()
        {
            var NewPhase = new Phase();
            NewPhase.Info = TransliteratePhases.Count.ToString();
            TransliteratePhases.Add(NewPhase);
            return NewPhase;
        }


        /// <summary>
        /// Apply the transliteration to the input string,
        /// </summary>
        /// <param name="InString">Input String</param>
        /// <returns>The converted result</returns>
        public string Convert(string InString)
        {
            return RunAllPhases(InString);
        }

        private string RunAllPhases(string InString)
        {
            var OutString = InString;

            foreach (var Phase in TransliteratePhases)
            {

                OutString = Phase.RunPhase(OutString);
            }

            return OutString;
        }

    }
}